### DISCUSSION TITLE
Clarification regarding new Trace segmentation code output

### DISCUSSION THREAD

**Discussion by Anurag19101996** [uw-ssec/MAWpy#74] [General]
Hello,

I had a small clarification regarding the new trace segmentation output. I ran the below attached file (100 users having orig_unc < 100m) for workflow 2, after commenting out update_stay_duration and wanted to ask few clarifications regarding the output:

1. If I am not wrong, the column Stay acts as the cluster identifier? Could you tell me what identifier is given for those traces which are part of no cluster? I initially thought it was 0, but since the Stay columns having 0 as the identifier have a stay_lat and stay_long associated, that wont be the case.
2. When I group the output of the new Trace Segmentation code by the column `Stay` and take the count of unique Stays, I get the value as 7101. When I group by (stay_lat, stay_long), the count shoots up to 10027. So I believe the difference in the counts are the *no* stay points or transient points - just adding on to my clarification on how to identify these points as I asked in point 1.
3. I ran Trace Segmentation in the older MAW code with spatial constraint 1 km and duration constraint as 300 s. I then selected the stay points (where stay_lat is not -1) and grouped this by (stay_lat, stay_long), which gave me 4275 clusters; different from the numbers in point 1 and point 2.

I may be going somewhere wrong in analyzing these numbers, could you take a look at this clarification please?
[100_GPS_users.csv](https://github.com/user-attachments/files/16385293/100_GPS_users.csv)

@anujsinha3 @carlosgjs @qzchen-uw @gracejia513

**Comment by anujsinha3** [comment]
Hi,
Please see the answers below:

1. The combination of columns (stay, user_id) are unique identifiers for traces with the same 'stay_lat' and 'stay_long'. Column 'stay' alone is not a unique identifier. This means that all traces that have the same (user_id, stay) will have the same value of 'stay_lat' and 'stay_long'

2. Can you confirm what you mean by 'no stay points' or 'transient_points'? Do they mean points that are not part of any cluster? If yes, do you want the values for (stay_lat, stay_long) to be represented as (-1, -1) or corresponding (orig_lat, orig_long) for these traces? (Since, in the new input file there will be only 4 cols - currently I have kept the default value as (orig_lat, orig_long), but I can change it accordingly)

3. There are multiple code issues/bugs in the old code, because of which the old code generates wrong outputs. A few examples are below:

- Date Parsing: The 'human_start_date' column in the input csv file is in string format: "1.90225E+11". In the old code to extract the date the first 6 characters of this string are used - that is equivalent to "1.9022". Since, this is not a valid date - the traces created per user per day are incorrect. [Code Reference](https://github.com/uw-ssec/MAWpy/blob/main/src/mawpy/TraceSegmentationClustering.py#L475).

- Incorrect loop condition: The old code also incorrectly terminates without checking the last trace for a day. In the [code](https://github.com/uw-ssec/MAWpy/blob/main/src/mawpy/TraceSegmentationClustering.py#L155), the conditional check should be "<=" instead of "<". Due to this, the last trace is always ignored and its values are never accounted for.

- Type Conversions: There are several cases of type casting from 'str' to 'float' and vice-versa. This generates precision loss. A few examples: [here](https://github.com/uw-ssec/MAWpy/blob/main/src/mawpy/TraceSegmentationClustering.py#L201) and [here](https://github.com/uw-ssec/MAWpy/blob/main/src/mawpy/TraceSegmentationClustering.py#L238)

These are a few examples that lead to incorrect output, hence I would not recommend using old code for benchmarking for accuracy.
Hope this helps.

**Comment by Anurag19101996** [comment]
Hi @anujsinha3, 

Thanks for the clarifications. Could you elaborate more on the precision loss statement above for type conversions?

We will get back to you on the first two points you mentioned regarding the old code. Meantime, as we discussed could you give us the code using the logic used in the [original algorithm](https://github.com/uw-ssec/MAWpy/blob/main/src/mawpy/TraceSegmentationClustering.py#L201). 

This would actually help us better understand if the logic change you proposed in the algorithm i.e. first checking spatial constraint then duration constraint as compared to the previous logic of first checking duration constraint then spatial constraint is causing any major differences.

**Reply by anujsinha3** [reply]
Python floats are formatted differently when using the str() function. Using str() on a float value and then vice-versa might lead to a change in the number of digits after decimal points in the represented values. And, sometimes this can also result in equality checks failure.

**Reply by anujsinha3** [reply]
https://github.com/anujsinha3/MAWpy/blob/feat-tsc-ic-usd/src/mawpy/steps/trace_segmentation_clustering_refactored_with_old_algorithm.py
This is the file (refactored with the old algorithm) in my repo (not on uw-ssec/main): 

As per my testing:

- Results Match: The trace segmentation clustering with refactored old logic vs the refactored new logic produced the same results for stay_lat and stay_long.
        Tested on spatial constraint 1 and duration_constraint: 0
        Tested on spatial_constraint 1 and duration_constraint: 600

- New Algorithmic Logic Faster: ~30% faster. This corresponds to approximately a 9-second difference for a file with  967796 traces. As measured on the M1 Pro 16Gb 6+2 core. The difference will increase even more for larger files.

I suggest that the ThinkLab team verify as well.

**Reply by anujsinha3** [reply]
Assume there are 15 traces for a user for a particular day. From here on - we will call this` list_of_traces`

According to the old algorithm, the code will check for a group of traces whose time difference is beyond the duration threshold. For this example, let's assume that the group is from index 0 to index 9 in the `list_of_traces`.

The timestamps for these 10 traces be: `timetamps_for_traces` = [21, 22, 25, 26, 28, 29, 35, 45, 55, 65] with time_constraint = 50

Now the old algorithm checks if the distance between any pair of traces violates the spatial constraint (n^2). **For our example, let's assume the pair to be the traces at indexes 7 and 9 which violates the spatial constraint.**

Since the spatial constraint is violated, the code will again start the calculation of the duration_constraint check and this time come up with index **1 to index 9.** 
Then it will again check spatial constraint - this will again break for the pair at indexes 7 and 9.

Then again, the code will again start the calculation of the duration_constraint check and this time come up with index 2 to index 9. 

Then it will again check spatial constraint - this will again break for the pair at index 7 and 9.

So on and so forth, for index 3-to-9, 4-to-9 .... up to 7-to-9

As you can see, we already know that the pair of indexes at 7 and 9 violate the spatial constraint, but the code is again and again checking for that pair, until the point where the range of index to be checked is from index 8-to-9.

Now to avoid this redundant calculation multiple times and utilize the knowledge of which index pair is responsible for breaking spatial constraint - I suggested flipping the order.

Hope, this helps.
